UpcomingMovies.Core/ViewModels/MoviesViewModel.cs

```diff
@@ -0,0 +1,268 @@
+
+            if (_currentStateContainer.MovieListType == MovieListType.Search)
+            {
+                SetMovieListToType(_previousStateContainer.MovieListType);
+            }
+            else
+            {
+                Movies.Clear();
+                _currentStateContainer.Movies.Clear();
+                _currentStateContainer.CurrentPage = 1;
+            }
+
+            await LoadMovies();
+            IsRefreshing = false;
+        }
+
+        MovieStateContainer GetMovieStateContainerForType(MovieListType movieListType)
+        {
+            if (_cachedMovies.ContainsKey(movieListType))
+                return _cachedMovies[movieListType];
+
+            var stateContainer = new MovieStateContainer { MovieListType = movieListType };
+            _cachedMovies.Add(movieListType, stateContainer);
+            return stateContainer;
+        }
+
+        /// <summary>
+        /// Controls what list should be showing in the "Movies" property, returning cached Movies and state.
+        /// </summary>
+        /// <param name="movieListType">Movie list type.</param>
+        void SetMovieListToType(MovieListType movieListType)
+        {
+            if (_currentStateContainer != null)
+            {
+                if (_currentStateContainer.MovieListType != MovieListType.Search)
+                    _previousStateContainer = _currentStateContainer;
+                else if (_currentStateContainer.MovieListType == movieListType)
+                    return;
+            }
+
+            _currentStateContainer = GetMovieStateContainerForType(movieListType);
+
+            Movies.Clear();
+            EmptySearchResults = false;
+            LoadFailed = false;
+            FailMessage = null;
+
+            if (_currentStateContainer.Movies.Count > 0) // Circumventing AiForms.CollectionView crash on Android
+                Movies.AddRange(_currentStateContainer.Movies);
+
+            switch (movieListType)
+            {
+                case MovieListType.Latest:
+                    Title = UiMessages.LatestMovies;
+                    break;
+                case MovieListType.NowPlaying:
+                    Title = UiMessages.NowPlaying;
+                    break;
+                case MovieListType.Popular:
+                    Title = UiMessages.PopularMovies;
+                    break;
+                case MovieListType.TopRated:
+                    Title = UiMessages.TopRatedMovies;
+                    break;
+                case MovieListType.Upcoming:
+                    Title = UiMessages.UpcomingMovies;
+                    break;
+                case MovieListType.Search:
+                    Title = UiMessages.SearchResults;
+                    break;
+            }
+        }
+
+        async Task ShowMovieDetailView(Movie movie)
+        {
+            IsNavigating = true;
+            await _navigationService.Navigate<MovieDetailViewModel, Movie>(movie);
+            IsNavigating = false;
```

Comment:
If the line above generates an exception `IsNavigating` will stuck in `true` state.

Reply:
Interesting, I did try setting the movie parameter to null and it still worked. I'm not sure when this would throw an exception, do you know?
I can put in a try/catch and in the finally block set IsNavigating to false but if navigating is throwing an exception the app is still not performing as intended, so this wouldn't address the underlying issue.

Reply:
Yeah, it's usually a good practice to wrap such pairing calls in try-finally.